### PR TITLE
Adding ADC support for stm323f303xc

### PR DESCRIPTION
**Pull request by cosmindanielradu19**
### Pull Request Overview

This pull request adds single sample ADC support for stm323f303xc

### Testing Strategy

This pull request was tested by using a stm323f3discovery kit

### TODO or Help Wanted

Continuous mode implementation troubles. The speed of interrupts is too high and can't send response with Data to the client.

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

File: boards/stm32f3discovery/src/main.rs [hunk 1]
```diff
@@ -497,6 +659,63 @@ pub unsafe fn reset_handler() {
...
+            // PE10 ADC3
+            &stm32f303xc::adc::Channel::Channel14,
+            // OA1 ADC1 (INTERNAL), PE11 ADC3
+            &stm32f303xc::adc::Channel::Channel15,
+            // TS ADC1 (INTERNAL), PE12 ADC3
+            &stm32f303xc::adc::Channel::Channel16,
+            // BT/2 ADC1 (INTERNAL), OA2 ADC2 (INTERNAL), OA3 ADC3 (INTERNAL), OA4 ADC4 (INTERNAL)
+            &stm32f303xc::adc::Channel::Channel17,
+            // VRI on ADC1, ADC2, ADC3 and ADC4 (INTERNALS)
+            &stm32f303xc::adc::Channel::Channel18,
```
Comment (hudson-ayers) [inline boards/stm32f3discovery/src/main.rs:702]: we should probably just pick a few ADC channels to include by default so that GPIOs are still available.

File: chips/stm32f303xc/src/adc.rs [hunk 2]
```diff
@@ -0,0 +1,740 @@
...
+//     // adc12: AdcSeparateRegisters,
+//     // _reserved0: [u32; 48],
+
+//     // adc34: AdcSeparateRegisters,
+//     // _reserved3: [u32; 109],
+
+//     // common: AdcCommonRegisters,
+
+// }
+
```
Comment (hudson-ayers) [inline chips/stm32f303xc/src/adc.rs:24]: is there any reason to keep these comments around?

File: chips/stm32f303xc/src/adc.rs [hunk 3]
```diff
@@ -0,0 +1,740 @@
...
+        DELAY OFFSET(8) NUMBITS(4) [],
+        // Dual ADC mode selection
+        DUAL OFFSET(0) NUMBITS(5) []
+    ],
+    // Common regular data register for dual mode
+    CDR [
+        // Regular data of the slave ADC
+        RDATA_SLV OFFSET(16) NUMBITS(16) [],
+        // Regular data of the master ADC
+        RDATA_MST OFFSET(0) NUMBITS(16) []
```
Comment (hudson-ayers) [inline chips/stm32f303xc/src/adc.rs:414]: These descriptive comments are awesome, but it would be great if you could make them doc comments so they show up in the docs (i.e. replace `//` --> `///`)

File: chips/stm32f303xc/src/adc.rs [hunk 4]
```diff
@@ -0,0 +1,740 @@
...
+pub static mut ADC1: Adc = Adc::new();
+
+impl Adc {
+    const fn new() -> Adc {
+        Adc {
+            registers: ADC1_BASE,
+            //rcc copy
+            common_registers: ADC12_COMMON_BASE,
+            clock: AdcClock(rcc::PeripheralClock::AHB(rcc::HCLK::ADC1)),
+            // clock34: AdcClock(rcc::PeripheralClock::AHB(rcc::HCLK::ADC34)),
```
Comment (hudson-ayers) [inline chips/stm32f303xc/src/adc.rs:527]: I noticed this clock is enabled in rcc.rs, is there a reason it is enabled there but ignored here?

File: chips/stm32f303xc/src/adc.rs [hunk 5]
```diff
@@ -0,0 +1,740 @@
...
+        // Wait for ADVRGEN to enable
+        // This needs to be synchronous because there is no interrupt signaling
+        // when ADVRGEN becomes enabled
+        for _i in 0..1000000 {
+            unsafe {
+                llvm_asm!(
+                "nop"
+            : : : : "volatile" );
+            }
+        }
```
Comment (hudson-ayers) [inline chips/stm32f303xc/src/adc.rs:554]: is there a bit we can check that will indicate when ADVRGEN has been enabled? Looping for a million cycles seems a shame if we can short circuit once it becomes ready, and it would be nice to not add `llvm_asm` to this crate if we don't have to.

File: chips/stm32f303xc/src/adc.rs [hunk 6]
```diff
@@ -0,0 +1,740 @@
...
+
+        // Enable the temperature sensor
+        self.common_registers.ccr.modify(CCR::TSEN::SET);
```
Comment (hudson-ayers) [inline chips/stm32f303xc/src/adc.rs:557]: What does this have to do with the ADC? Is it used invisibly by the hardware to correct for temperature affects somehow?

File: chips/stm32f303xc/src/adc.rs [hunk 7]
```diff
@@ -0,0 +1,740 @@
...
+        self.registers.ier.modify(IER::ADRDYIE::SET);
+
+        // Clear registers
+        self.registers.isr.modify(ISR::ADRDY::CLEAR);
+        self.registers.cr.modify(CR::ADEN::CLEAR);
+        self.registers.cr.modify(CR::ADCALDIF::CLEAR);
+        self.registers.cr.modify(CR::ADCAL::SET);
+
+        // Wait for calibration
+        while self.registers.cr.is_set(CR::ADCAL) {}
```
Comment (hudson-ayers) [inline chips/stm32f303xc/src/adc.rs:569]: Do you have a sense of how many cycles this wait should take? As a general rule, we favor asynchronous designs unless the wait is expected to be very small.

File: chips/stm32f303xc/src/adc.rs [hunk 8]
```diff
@@ -0,0 +1,740 @@
...
+            // Clear interrupt
+            self.registers.ier.modify(IER::OVRIE::CLEAR);
+            self.registers.isr.modify(ISR::OVR::SET);
+        }
+    }
+
+    pub fn set_client<C: EverythingClient>(&self, client: &'static C) {
+        self.client.set(client);
+    }
+    // 12
```
Comment (hudson-ayers) [inline chips/stm32f303xc/src/adc.rs:621]:
```suggestion

```

File: chips/stm32f303xc/src/adc.rs [hunk 9]
```diff
@@ -0,0 +1,740 @@
...
+        _length1: usize,
+        _buffer2: &'static mut [u16],
+        _length2: usize,
+    ) -> (
+        ReturnCode,
+        Option<&'static mut [u16]>,
+        Option<&'static mut [u16]>,
+    ) {
+        (ReturnCode::ENOSUPPORT, None, None)
+    }
```
Comment (hudson-ayers) [inline chips/stm32f303xc/src/adc.rs:714]: If high speed is not supported, it seems like it makes more sense to simply not implement the trait. Is there something about how the ADC capsule is written that makes doing so not ergonomic?

File: chips/stm32f303xc/src/adc.rs (hunk 5 above)
Comment (cosmindanielradu19) [reply chips/stm32f303xc/src/adc.rs:554]: No, it just needs a wait time for it to be enabled and to continue the ADC powering on

File: chips/stm32f303xc/src/adc.rs (hunk 4 above)
Comment (cosmindanielradu19) [reply chips/stm32f303xc/src/adc.rs:527]: At the moment just ADC12 is used and tested, ADC34 is on progress, just need to test a bit more

File: chips/stm32f303xc/src/adc.rs (hunk 7 above)
Comment (cosmindanielradu19) [reply chips/stm32f303xc/src/adc.rs:569]: No, there is not a fixed amount of cycles needed that's why I chose busy waiting for this to wait the completion

File: chips/stm32f303xc/src/adc.rs (hunk 9 above)
Comment (cosmindanielradu19) [reply chips/stm32f303xc/src/adc.rs:714]: Couldn't compile ADC and integrate it in the kernel without high speed Hil implementation

File: chips/stm32f303xc/src/adc.rs (hunk 6 above)
Comment (cosmindanielradu19) [reply chips/stm32f303xc/src/adc.rs:557]: It will be enabled on temperature capsule, it was just for testing. I will do this as soon as I get ADC mux working

File: chips/stm32f303xc/src/adc.rs (hunk 5 above)
Comment (cosmindanielradu19) [reply chips/stm32f303xc/src/adc.rs:554]: This is the link to the reference manual for stm32f3xx https://www.st.com/resource/en/reference_manual/dm00043574-stm32f303xb-c-d-e-stm32f303x6-8-stm32f328x8-stm32f358xc-stm32f398xe-advanced-arm-based-mcus-stmicroelectronics.pdf and on page 316 it is described how to enable ADVREGEN. Am I missing anything? I’m open to any ideas.

File: chips/stm32f303xc/src/adc.rs (hunk 7 above)
Comment (cosmindanielradu19) [reply chips/stm32f303xc/src/adc.rs:569]: This is the link to the reference manual for stm32f3xx https://www.st.com/resource/en/reference_manual/dm00043574-stm32f303xb-c-d-e-stm32f303x6-8-stm32f328x8-stm32f358xc-stm32f398xe-advanced-arm-based-mcus-stmicroelectronics.pdf and on page 317 it is described how to do the calibration process. It says that the ADCAL bit will be cleared by the hardware as soon as the process is completed.

File: chips/stm32f303xc/src/adc.rs (hunk 9 above)
Comment (cosmindanielradu19) [reply chips/stm32f303xc/src/adc.rs:714]: If we can split the ADC capsule in two different parts, more exactly ADC and High Speed ADC, this would work, otherwise the implementation for High Speed is needed for the capsule to be compiled alongside the kernel.

File: boards/stm32f3discovery/src/main.rs (hunk 1 above)
Comment (cosmindanielradu19) [reply boards/stm32f3discovery/src/main.rs:702]: Ok, I will remove some of the ADC pins and leave them as GPIOs and I will come back with an update

File: chips/stm32f303xc/src/adc.rs (hunk 5 above)
Comment (hudson-ayers) [reply chips/stm32f303xc/src/adc.rs:554]: That seems to indicate the worst case delay is 10 microseconds. Also, my understanding is that the maximum core clock speed is 72 MHz on this chip, so I think the worst case number of cycles should be 720. Can you replace 1000000 with 720, and add a comment explaining why that number is used? 

Ideally we would have a safe wait primitive like that described in #1552 , but for now I think limiting this to 720 cycles should reflect the spirit of that.

File: chips/stm32f303xc/src/adc.rs (hunk 6 above)
Comment (hudson-ayers) [reply chips/stm32f303xc/src/adc.rs:557]: Should this not be enabled in the temperature capsule then?

File: boards/stm32f3discovery/.vscode/c_cpp_properties.json [hunk 10]
```diff
@@ -0,0 +1,16 @@
+{
+    "configurations": [
+        {
+            "name": "Linux",
+            "includePath": [
+                "${workspaceFolder}/**"
+            ],
+            "defines": [],
+            "compilerPath": "/usr/bin/gcc",
```
Comment (alistair23) [inline boards/stm32f3discovery/.vscode/c_cpp_properties.json:9]: This probably shouldn't be included in the Tock repo

File: boards/stm32f3discovery/src/main.rs [hunk 11]
```diff
@@ -358,94 +436,103 @@ pub unsafe fn reset_handler() {
...
+            86 => stm32f303xc::gpio::PinId::PC09.get_pin().as_ref().unwrap()
         ),
     )
     .finalize(components::gpio_component_buf!(
         stm32f303xc::gpio::Pin<'static>
     ));
 
+    //start adc12
+    stm32f303xc::adc::ADC1.enable();
+
```
Comment (bradjc) [inline boards/stm32f3discovery/src/main.rs:535]: Can you move this down to the other ADC setup code?

File: chips/stm32f303xc/src/adc.rs [hunk 12]
```diff
@@ -0,0 +1,726 @@
+use crate::rcc;
```
Comment (bradjc) [inline chips/stm32f303xc/src/adc.rs:1]:
```suggestion
//! Analog to Digital Converter Peripheral

use crate::rcc;
```

File: chips/stm32f303xc/src/adc.rs (hunk 9 above)
Comment (bradjc) [reply chips/stm32f303xc/src/adc.rs:700]: Hmm ok that makes sense. I'm not sure about providing a fake interface to userspace, the ADC capsule still isn't supported on the stm32 at this point. I can imagine being on the other side as a developer and thinking ADC will work and then learning why all of the syscalls are returning errors.

Comment (cosmindanielradu19) [reply chips/stm32f303xc/src/adc.rs:702]: I still suggest splitting the ADC capsule in two as this will allow an easier implementation of a virtual ADC

File: boards/stm32f3discovery/src/main.rs [hunk 13]
```diff
@@ -497,6 +581,40 @@ pub unsafe fn reset_handler() {
     let ninedof = components::ninedof::NineDofComponent::new(board_kernel)
         .finalize(components::ninedof_component_helper!(l3gd20, lsm303dlhc));
 
+    //start adc12
+    stm32f303xc::adc::ADC1.enable();
```
Comment (brghena) [inline boards/stm32f3discovery/src/main.rs:585]: Why enable the ADC here instead of within the `sample()` call? The general idea when the ADC HIL was written was that calls to `sample()` should enable, configure, and sample the ADC so that a developer can't ever forget to do part of that and have the ADC in an indeterminate state.

File: chips/stm32f303xc/src/adc.rs (hunk 9 above)
Comment (brghena) [reply chips/stm32f303xc/src/adc.rs:702]: I think that the ADC capsule should have been (and still should be) split in two. That didn't happen at the time because of the complexity of implementing virtualization to share one ADC between two interfaces, which remains true today.

I think in lieu of more immediate work on the ADC capsule side, this ENOSUPPORT implementation is a reasonable choice. I would follow this up with a comment in main.rs by the ADC configuration that the highspeed ADC trait is not actually implemented for now, and when you fill out https://github.com/tock/tock/blob/master/chips/README.md#hil-support, I would not mark highspeed ADC as implemented.

File: chips/stm32f303xc/src/rcc.rs [hunk 14]
```diff
@@ -594,6 +595,36 @@ impl Rcc {
...
     }
+
+    // ADC12 clock
+
+    fn is_enabled_adc12_clock(&self) -> bool {
+        self.registers.ahbenr.is_set(AHBENR::ADC12EN)
+    }
+
+    fn enable_adc12_clock(&self) {
+        debug!("enable adc12 clock");
```
Comment (bradjc) [inline chips/stm32f303xc/src/rcc.rs:606]:
```suggestion
```

File: chips/stm32f303xc/src/adc.rs [hunk 15]
```diff
@@ -0,0 +1,752 @@
...
+            common_registers: ADC12_COMMON_BASE,
+            clock: AdcClock(rcc::PeripheralClock::AHB(rcc::HCLK::ADC1)),
+            // clock34: AdcClock(rcc::PeripheralClock::AHB(rcc::HCLK::ADC34)),
+            status: Cell::new(ADCStatus::Off),
+            client: OptionalCell::empty(),
+        }
+    }
+
+    pub fn enable(&self) {
+        debug!("ENABLE");
```
Comment (bradjc) [inline chips/stm32f303xc/src/adc.rs:526]:
```suggestion
```

File: chips/stm32f303xc/src/adc.rs [hunk 16]
```diff
@@ -0,0 +1,752 @@
+//! Analog to Digital Converter Peripheral
+
+use crate::rcc;
+use core::cell::Cell;
+use kernel::common::cells::OptionalCell;
+use kernel::common::registers::{register_bitfields, ReadOnly, ReadWrite};
+use kernel::common::StaticRef;
+use kernel::debug;
```
Comment (bradjc) [inline chips/stm32f303xc/src/adc.rs:8]:
```suggestion
```

**Comment by bradjc** [comment]
@hudson-ayers
